migration/ubuntu/20-to-24.md

```diff
@@ -0,0 +1,242 @@
+
+  ```bash
+  kubectl edit configmap -n kube-public cluster-info
+  ```
+
+- Set internal IP of the first new control plane node into the ConfigMap `kube-system/kubeadm-config` under the key `controlPlaneEndpoint`
+
+  Replace any old control plane hostnames and IP with the new control plane in the same ConfigMap under the key `certSANs`.
+
+  ```bash
+  kubectl edit configmap -n kube-system kubeadm-config
+  ```
+
+- Update DNS records to point to the new control plane nodes
+
+> [!NOTE]
+> This is only needed if you are using DNS and are pointing that to the IPs of the control plane nodes.
+> It could be updating `dns.json` and applying if using AWS
+
+- Update loadbalancer to use the endpoints of the new control plane nodes
+
+> [!NOTE]
+> This is only needed if you are using a loadbalancer in front of the API server.
+
+- Update control plane order using Kubespray:
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray apply <sc|wc> --tags=master,nginx -b -e=ignore_assert_errors=true
+  popd
+  ```
+
+- Remove old control plane nodes using Kubespray:
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray remove-node <sc|wc> <old-control-plane-nodes,...>
+  popd
+  ```
+
+> [!NOTE]
+> This will likely fail at some point in one or two ways:
+>
+> - If it fails early for some and succeed for others then remove the removed nodes from inventory and retry.
+> - If it fails late for some with timeouts then ignore.
+
+- Check that old etcd members are removed:
+
+  ```bash
+  HOSTNAME="$(kubectl get node -l node-role.kubernetes.io/control-plane= -ojsonpath='{.items[0].metadata.name}')"
+  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
+    etcdctl --endpoints=https://127.0.0.1:2379 \
+            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
+            --cert=/etc/kubernetes/ssl/etcd/server.crt \
+            --key=/etc/kubernetes/ssl/etcd/server.key \
+            member list
+  ```
+
+- If there are old nodes left, remove them using the `uuid` from the previous command (The id in the first column of each row)
+
+  ```bash
+  HOSTNAME="$(kubectl get node -l node-role.kubernetes.io/control-plane= -ojsonpath='{.items[0].metadata.name}')"
+  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
+    etcdctl --endpoints=https://127.0.0.1:2379 \
+            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
+            --cert=/etc/kubernetes/ssl/etcd/server.crt \
+            --key=/etc/kubernetes/ssl/etcd/server.key \
+            member remove <uuid>
+  ```
+
+- Check that old control plane nodes are removed:
+
+  ```bash
+  kubectl get nodes
+  ```
+
+- If there are any old nodes left, delete them
+
+  ```bash
+  kubectl delete nodes <node>,...
+  ```
+
+- Change back to using dynamic inventory
+
+  ```bash
+  pushd "${CK8S_CONFIG_PATH}"
+  cp -r <sc|wc>-config-backup/inventory.ini <sc|wc>-config/inventory.ini
```

Comment:
Do we need recursive here?

Reply:
https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/9bf216c10273ee3ba15810b029a3366a796a7e94